### PR TITLE
 Switching from sstream to c string formatting to fix ros arg issue

### LINKED ISSUE #540: TF2 throws rclcpp::exceptions::RCLInvalidROSArgsError when Pytorch libraries are linked
Question originally asked here:
https://answers.ros.org/question/403755/tf2-throws-rclcppexceptionsrclinvalidrosargserror-when-pytorch-libraries-are-linked/?answer=403759#post-id-403759

## Bug report

**Required Info:**

- Operating System:
  - Ubuntu 20.04
- Installation type:
  - Binaries via apt
- Version or commit hash:
  - 1186ee1e1be1dfe3691f31ba7f13662b1f1e1b91
- DDS implementation:
  - Fast-RTPS
- Client library (if applicable):
  - rclcpp

#### Steps to reproduce issue
Add these lines to CMakeLists.txt file that also links TF2 libraries
```
find_package(Torch REQUIRED)

target_link_libraries(${PROJECT_NAME} ${TORCH_LIBRARIES})
```

#### Expected behavior
TransformListener initializes and gives valid transform data.

#### Actual behavior
A run time error is thrown:
```
what():  failed to parse arguments: Couldn't parse remap rule: '-r __node:=transform_listener_impl_563,a2c,cfa,cc0'. Error: Expected lexeme type (1) not found, search ended at index 35, at /tmp/binarydeb/ros-galactic-rcl-3.1.3/src/rcl/lexer_lookahead.c:233, at /tmp/binarydeb/ros-galactic-rcl-3.1.3/src/rcl/arguments.c:371
```

#### Additional information
Removing the pytorch library link causes this issue to disappear. For some reason, when torch libraries are linked, the behavior of string stream formatting changes. When strings are passed via the `<<` operator to `std::hex`, commas are inserted. See this line: https://github.com/ros2/geometry2/blob/2687a3543bf61481b06dcf37839c7e92bed85d2e/tf2_ros/src/transform_listener.cpp#L48

So when torch isn't linked, the string produced is `transform_listener_impl_563a2ccfacc0`. With torch linked, the string becomes `transform_listener_impl_563,a2c,cfa,cc0` causing the rcl arg parser to freak out.

#### Suggested fix
Avoid using std::hex and string streams. I have a fix on this fork: https://github.com/frc-88/geometry2/commit/0adfa8f83a05f77527be8b3eab2c8aaa5630adc9

### PR DESCRIPTION
**Pull request by clalancette**
See https://github.com/ros2/geometry2/issues/540 for issue details.

TL;DR, linking certain libraries overrides sstream behavior. In this case, random commas were inserted causing rcl arg parser to freak out and throw a runtime error.

I've switched this line to use sprintf to generate a hex string instead of sstream and std::hex.

Note that this is a replacement for #542 , containing exactly the same code but rebased onto the latest and with a fix for a Windows warning.

### CODE REVIEW

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17520/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/12055/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/18074/)

**Comment by iuhilnehc-ynos** [comment]
LGTM. The following patch could be another solution.

```patch
   std::stringstream sstream;
+  sstream.imbue(std::locale::classic());
   sstream << "transform_listener_impl_" << std::hex << reinterpret_cast<size_t>(this);
```

**Comment by clalancette** [comment]
> LGTM. The following patch could be another solution.

Hm, true.  I'm going to go ahead with the current solution since I know it was tested to work in #542 .  Thanks for the review, merging.
